### ISSUE TITLE
BUG: Changing dtype from float64 to complex64 or 128 causes unexpected behaviour

### ISSUE THREAD

**Issue by Wikstahl** [numpy/numpy#20715]
### Describe the issue:

I was naive trying to change `dtype` of an array from `float64` to `complex64` by simply overwriting the `dtype` as `arr.dtype = np.complex64`. However this caused all elements in my array to either become super big or super small. I now know that this is the wrong way to change the `dtype` of an array. However, I think that an error-message or warning should have been in place.

### Reproduce the code example:

```python
import numpy as np
arr = np.random.rand(5,5)
print(arr)
arr.dtype = np.complex64
print(arr)
```

### NumPy/Python version information:

1.21.5 3.9.9 (main, Nov 21 2021, 03:23:44) 
[Clang 13.0.0 (clang-1300.0.29.3)]

**Comment by matthew-brett** [comment]
It's not completely unreasonable to change the Dtype, but as you've discovered, the Dtype tells Numpy how to interpret the memory in the array.   So - what warning would you recommend?   For example, I could imagine "Warning: you are changing the interpretation of memory in the array, but not changing the underlying memory values" - but that would be annoying for someone who really was trying to change the interpretation.

**Comment by mattip** [comment]
In the experimental array API the dtype is readonly, so perhaps you could try using that instead:
```python
>>> import numpy.array_api as xp
>>> import numpy as np
>>> arr = xp.asarray(np.random.rand(5,5))
arr.dtype = np.complex64
>>> arr.dtype = np.complex64
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: can't set attribute
```

NumPy is unlikely to change this, but I could imagine a softer "DoAsISay" version of NumPy that implements the array API and would do things like warn on dangerous behaviour, error on int overflow. Using the Array API would already prevent object dtype from being created, (but also would remove string, byte, and struct dtypes).

**Comment by charris** [comment]
We did make a number of those read only back in the day. ISTR that dtype was left available for compatibility reasons.

**Comment by seberg** [comment]
Yeah, dtype _should_ be readonly definitely.  But I am not sure how much noise it would make to change.  I would be +1 for a deprecation, with the caveat that it should not create too much noise downstream.

IIRC one issue with these was that masked arrays did make use of some of these shadier "mutable container" APIs.

**Comment by charris** [comment]
> But I am not sure how much noise it would make to change

There were downstream programs that were using it, but that was maybe 8-10 years ago.

**Comment by mhvk** [comment]
Trying to move from setting `dtype` and `shape` to using `.view` and `.reshape` would seem sensible...

But as a simpler solution, would it be possible to mention in the docstring of `np.ndarray.dtype` what setting does, and why a `.view` is preferred? Or, even better, could one add a useful docstring for `array_instance.dtype`?

**Comment by seberg** [comment]
You are right, we should at least start with warning about it.  I don't think anything really picks up the docstring for instances, but we can put it on the property (and have it). Opened gh-20730 to at least add a scary warning :).
